cupyx/jit/_cuda_types.py

```diff
@@ -1,84 +1,100 @@
 
+if TYPE_CHECKING:
+    from cupyx.jit._internal_types import Data
+
 
 # Base class for cuda types.
 class TypeBase:
 
-    def __str__(self):
+    def __str__(self) -> str:
         raise NotImplementedError
 
-    def declvar(self, x, init):
+    def declvar(self, x: str, init: 'Data') -> str:
         if init is None:
             return f'{self} {x}'
         return f'{self} {x} = {init.code}'
 
-    def assign(self, var, value):
+    def assign(self, var: 'Data', value: 'Data') -> str:
         return f'{var.code} = {value.code}'
 
 
 class Void(TypeBase):
 
-    def __init__(self):
+    def __init__(self) -> None:
         pass
 
-    def __str__(self):
+    def __str__(self) -> str:
         return 'void'
 
 
 class Scalar(TypeBase):
 
-    def __init__(self, dtype):
+    def __init__(self, dtype: npt.DTypeLike) -> None:
         self.dtype = numpy.dtype(dtype)
 
-    def __str__(self):
+    def __str__(self) -> str:
         dtype = self.dtype
         if dtype == numpy.float16:
             # For the performance
             dtype = numpy.dtype('float32')
         return get_typename(dtype)
 
-    def __eq__(self, other):
+    def __eq__(self, other: TypeBase) -> bool:  # type: ignore[override]
         return isinstance(other, Scalar) and self.dtype == other.dtype
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         return hash(self.dtype)
 
 
 class PtrDiff(Scalar):
-    def __init__(self):
+
+    def __init__(self) -> None:
         super().__init__('q')
 
-    def __str__(self):
+    def __str__(self) -> str:
         return 'ptrdiff_t'
 
 
 class ArrayBase(TypeBase):
 
-    def __init__(self, child_type: TypeBase, ndim: int):
+    def __init__(self, child_type: TypeBase, ndim: int) -> None:
         assert isinstance(child_type, TypeBase)
         self.child_type = child_type
         self.ndim = ndim
 
 
 class CArray(ArrayBase):
-    def __init__(self, dtype, ndim, is_c_contiguous, index_32_bits):
-        self.dtype = dtype
+
+    def __init__(
+            self,
+            dtype: npt.DTypeLike,
+            ndim: int,
+            is_c_contiguous: bool,
+            index_32_bits: bool,
+    ) -> None:
+        self.dtype = numpy.dtype(dtype)
         self._c_contiguous = is_c_contiguous
         self._index_32_bits = index_32_bits
         super().__init__(Scalar(dtype), ndim)
 
     @classmethod
-    def from_ndarray(cls, x):
+    def from_ndarray(cls, x: cupy.ndarray) -> 'CArray':
         return CArray(x.dtype, x.ndim, x._c_contiguous, x._index_32_bits)
 
-    def __str__(self):
+    def __str__(self) -> str:
         ctype = get_typename(self.dtype)
         c_contiguous = get_cuda_code_from_constant(self._c_contiguous, bool_)
         index_32_bits = get_cuda_code_from_constant(self._index_32_bits, bool_)
         return f'CArray<{ctype}, {self.ndim}, {c_contiguous}, {index_32_bits}>'
 
-    def __eq__(self, other):
+    def __eq__(self, other: TypeBase) -> bool:  # type: ignore[override]
```

Comment:
I think `other` should be `object`.
https://stackoverflow.com/questions/37557411/why-does-defining-the-argument-types-for-eq-throw-a-mypy-type-error